### ISSUE TITLE
indexing problem

### ISSUE THREAD

**Issue by emauryg** [mlverse/torch#691]
Sometimes after indexing into a tensor, the index changes to 0-based rather than 1-based, creating problems if the same indexing variable is used 

See below
![image](https://user-images.githubusercontent.com/29872305/134076918-50a0544f-dc64-4699-b506-e5f6d04e06a7.png)

**Comment by mohamed-180** [comment]
Strange behavior of function `[`  on tensors , you can use work around by using slicing like :
```
a <- torch_randn(c(6,4))
b <- c(1:3)
a[b]
b

>torch_tensor
 0.1952 -0.7868 -1.3279 -0.6170
-0.1944 -0.3089  0.1955  0.3347
 0.4781  0.3171  0.4963  0.9246
[ CPUFloatType{3,4} ]
1 . 2 . 3
```

**Comment by emauryg** [comment]
is this something that can be fixed on the next version release? For reference, device is a cuda device on my example above.

**Comment by dfalbel** [comment]
This is definetely a bug, we should make a copy of the `slice` vector here:

https://github.com/mlverse/torch/blob/master/src/indexing.cpp#L147

Thanks for reporting! We will make sure to fix this in the next release. 
As a workaround you can do something like:

```
x[torch_tensor(as.integer(y))]
```

So a copy is always created before indexing.

**Comment by emauryg** [comment]
Oh thanks that would be great!

**Comment by mohamed-180** [comment]
Can we change :
https://github.com/mlverse/torch/blob/1344c3aae4d91db5068608cae44891be3d99d966/src/indexing.cpp#L145-L168

To some thing like : 
```
  Rcpp::NumericVector v(slice.length());
  for (int j = 0; j < slice.size(); j++)
  {
    if (slice[j] > 0)
    {
      v[j] = slice[j] - 1; // make it 0-based.
    }
    else if (slice[j] == 0)
    {
      Rcpp::stop("Indexing in R is 1-based and found a 0.");
    }
  }
  
  // Create the integer Tensor
  XPtrTorchTensorOptions options = lantern_TensorOptions();
  options = lantern_TensorOptions_dtype(options.get(), XPtrTorchDtype(lantern_Dtype_int64()).get());
  std::vector<int64_t> dim = {LENGTH(slice)};
  
  XPtrTorchTensor tensor = cpp_torch_tensor(v, dim, options, false, false);
  
  lantern_TensorIndex_append_tensor(index.get(), tensor.get());
}
```
To avoid change the slice object

**Comment by dfalbel** [comment]
Yes @mohamed-180 ! This sounds like a good solution. Do you want to file a PR?

**Comment by emauryg** [comment]
I think there is another issue now, 

```
a <- torch_randn(c(3,4,2))
c <- torch_randn(c(3,4,2))
b <- c(1,2,4)
a[..,torch_tensor(as.integer(b)),]
c[..,torch_tensor(as.integer(b)),]
b
```

`Error in Tensor_slice(tensor$ptr, environment(), drop = drop, mask = .d): index 3 is out of bounds for dimension 0 with size 2
Exception raised from get at aten/src/ATen/native/cpu/IndexKernel.cpp.AVX2.cpp:42 (most recent call first)`

The error does not occur if we do not use the `..` annotation.

**Comment by dominikj2** [comment]
Hi @mohamed-180, 
does above solved problem #691 also solve the second issue presented by @emauryg. I am having a similar problem with` ..` annotation.
Thanks!

**Comment by dfalbel** [comment]
Hi @dominikj2 , no not yet. I'll actually open another issue for this so it's easier to find out.
